core/trino-main/src/main/java/io/trino/server/remotetask/HttpRemoteTask.java

```diff
@@ -579,73 +584,75 @@ private void triggerUpdate()
         }
     }
 
-    private synchronized void sendUpdate()
+    private void sendUpdate()
```

Comment:
If we remove `synchronized` here and from `public void setOutputBuffers(OutputBuffers newOutputBuffers)`
we change the semantic, the methods `sendUpdate` and `setOuputBuffer` are not mutually exclusive. 
In a result we have a situation when the state of "reported" (published by `sendUpdate`) is internally consistent. 

Especially, we could mix values `sendPlan` and `sentDynamicFiltersVersion` in `sendUpdate` from the last one response and from the previous one.

Reply:
Yeah - concerns look valid.
@sopel39 what exactly do you want to fix here? Is there a performance bottle neck this change addresses, and is that tested it actually helps. 
If benefit is small I would just leave current simple synchronization. If we really need that the change need to be fine crafted so it is easy to follow (this change is not).

Reply:
> If we remove synchronized here and from success we change the semantic, the methods sendUpdate and sucess are not mutually exclusive.

`sendUpdate` and `sucess` were never mutually exclusive. However, they still synchronize on `httpClient`, so appropriate fields are guarded.

> In a result we have a situation when the state of "reported" request (published by sendUpdate) is internally inconsistent.

I don't understand that

> Especially, we could mix values sendPlan and sentDynamicFiltersVersion in sendUpdate from the last one response and from the previous one.

It doesn't matter. Both `sentDynamicFiltersVersion` and `sendPlan` are idempotent. `sentDynamicFiltersVersion` never goes down.

> @sopel39 what exactly do you want to fix here? Is there a performance bottle neck this change addresses, and is that tested it actually helps.

Take a look at https://github.com/trinodb/trino/commit/583fcabd51a3757bd4e4ccb4401ebc59e2b5dafc. This issue caused serious regression from release to release. However, that was not a proper fix (rather some "lucky" relaxation of constraints) as all methods are still synchronized on `this` (including expensive `sendUpdate`). This PR splits splits `sendUpdate` synchronization fully

Reply:
> sendUpdate and sucess were never mutually exclusive. However, they still synchronize on httpClient, so appropriate fields are guarded.
 
`sentDynamicFiltersVersion` and `sendPlan` are not under `httpClient` lock in the `success` method. 

> I don't understand that
> It doesn't matter. Both sentDynamicFiltersVersion and sendPlan are idempotent. sentDynamicFiltersVersion counter never goes down.

What do you mean they are idempotent? `value.isNeedsPlan` from response (look to the `success` method) cannot be different from `HttpRemoteTask.sendPlan` that is read in `updateTask` from another thread?

Reply:
> What do you mean they are idempotent?

It doesn't matter if coordinator sends some DF or plan twice (and `HttpRemoteTask.sendPlan` is true by default). I've changed the code to make sure `sentDynamicFiltersVersion` and `sendPlan` is updated before next request is started (`currentRequest = null`)